ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala

```diff
@@ -0,0 +1,259 @@
+import com.daml.platform.store.backend.common.UserManagementStorageBackendTemplate
+import com.daml.platform.usermanagement.PersistentUserManagementStore.TooManyUserRightsRuntimeException
+
+import scala.concurrent.{ExecutionContext, Future}
+
+object PersistentUserManagementStore {
+
+  /** Intended to be thrown within a DB transaction to abort it.
+    * The resulting failed future will get mapped to a successful future containing scala.util.Left
+    */
+  final case class TooManyUserRightsRuntimeException(userId: Ref.UserId) extends RuntimeException
+
+  def cached(dbDispatcher: DbDispatcher, metrics: Metrics)(implicit
+      executionContext: ExecutionContext
+  ): UserManagementStore = {
+    new CachedUserManagementStore(
+      delegate = new PersistentUserManagementStore(
+        dbDispatcher = dbDispatcher,
+        metrics = metrics,
+        maxNumberOfUserRightsPerUserLimit = ApiVersionService.MaxNumberOfUserRightsPerUser,
+      ),
+      expiryAfterWriteInSeconds = CachedUserManagementStore.ExpiryAfterWriteInSeconds,
+    )
+  }
+
+  def cachedAuthorizer(userManagementStore: UserManagementStore)(implicit
+      executionContext: ExecutionContext,
+      loggingContext: LoggingContext,
+  ): UserManagementBasedAuthorizer = {
+    new CachedUserManagementBasedAuthorizer(
+      delegate = new UserManagementBasedAuthorizerImpl(
+        userManagementStore = userManagementStore
+      )(
+        executionContext = executionContext,
+        loggingContext = loggingContext,
+      ),
+      expiryAfterWriteInSeconds = CachedUserManagementStore.ExpiryAfterWriteInSeconds,
+    )
+  }
+}
+
+class PersistentUserManagementStore(
+    dbDispatcher: DbDispatcher,
+    metrics: Metrics,
+    createAdminUser: Boolean = true,
+    maxNumberOfUserRightsPerUserLimit: Int,
+) extends UserManagementStore {
+
+  private val backend: UserManagementStorageBackend = UserManagementStorageBackendTemplate
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  implicit private val loggingContext: LoggingContext = LoggingContext.newLoggingContext(identity)
+
+  // TODO participant user management: This is probably not the best place/time to create admin user
+  if (createAdminUser)
+    createUser(
+      InMemoryUserManagementStore.AdminUser.user,
+      InMemoryUserManagementStore.AdminUser.rights,
+    )
+
+  override def createUser(
+      user: domain.User,
+      rights: Set[domain.UserRight],
+  ): Future[Result[Unit]] = {
+    inTransaction { implicit connection: Connection =>
+      withoutUser(user.id) {
+        if (rights.size > maxNumberOfUserRightsPerUserLimit) {
+          throw TooManyUserRightsRuntimeException(user.id)
+        } else {
+          val nowMicros = epochMicroSeconds()
+          val internalId = backend.createUser(user, createdAt = nowMicros)(connection)
+          rights.foreach(right =>
+            backend.addUserRight(internalId = internalId, right = right, grantedAt = nowMicros)(
```

Comment:
This call may fail due to uniqueness constraint violation. It needs to be handled.

Reply:
In this particular place we should be immune from user rights uniqueness violation:
1) `rights: Set[domain.UserRight]` doesn't contain duplicates
2) we've just created the user in the same transaction so no other transaction can see it yet, thus no other transaction can add a duplicated user right

Reply:
TODO: write ledger api tests that handles exception raised when race cond occurs